faq.md

```diff
@@ -1,45 +1,634 @@
+    
+    public class MyExampleWindow : GameWindow
+    {
+        public MyExampleWindow(GameWindowSettings gameWindowSettings,
+                              NativeWindowSettings nativeWindowSettings)
+            : base(gameWindowSettings, nativeWindowSettings)
+        {
+        }
+        
+        protected override void OnUpdateFrame(FrameEventArgs e)
+        {
+			// This gets called every 1/60 of a second.
+            if (KeyboardState.IsKeyDown(Keys.Escape))
+                Close();
+            
+            base.OnUpdateFrame(e);
+        }
+        
+        protected override void OnRenderFrame(FrameEventArgs e)
+        {
+            // Show that we can use OpenGL: Clear the window to cornflower blue.
+            GL.ClearColor(0.39f, 0.58f, 0.93f, 1.0f);
+            GL.Clear(ClearBufferMask.ColorBufferBit);
+        }
+    }
+}
+```
+
+More sophisticated uses will load textures and shaders and models during the `OnLoad()` handler, clean up during the `OnUnload()` handler, and do a lot more than just wait for the user to press `Escape` to close the window — but the code above is enough to get you started.
+
+(Creating a window in OpenTK 3 requires a similar amount of code, but it looks a bit different from this example.  New projects should use OpenTK 4+.)
+
+### How do I call OpenGL/CL/AL functions?
+
+In your code, you can invoke the OpenGL/CL/AL functions directly using the static classes that OpenTK exposes.  OpenTK provides all of the same functions that the underlying native libraries provide, just with the initial `gl*` or `cl*` or `al*` prefix turned into a static class name.  So, for example:
+
+- In C you’d call `int error = glGetError();`.  In C# you’d write `ErrorCode error = GL.GetError();`.
+- In C you’d call `glBindVertexArray(handle);`.  In C# you’d write `GL.BindVertexArray(handle);`.
+- In C you’d write `glCreateShader(GL_FRAGMENT_SHADER);`.  In C# you’d write `GL.CreateShader(ShaderType.FragmentShader)`.
+
+As you can see, the programming models are nearly identical, by design:  If you know how to use OpenGL/OpenCL/OpenAL, you know how to use OpenTK.
+
+**Arrays and pointers**
+
+Passing arrays from .NET to OpenGL/CL/AL (or to any C library) is a complex topic, but OpenTK has built-in support for it:  See the section below on “[How do I pass pointers and arrays?](#How-do-I-pass-pointers-and-arrays)”.
+
+**Constant symbols**
+
+Note in the third example that the constants change syntax slightly:  They aren’t global `#define` directives as in C, but instead have some minor alterations to make them more .NET-friendly:
+
+- The constants are grouped into strong `enum` types like `ErrorCode` and `ShaderType` — they’re not just a big pile of unrelated values like they are in C.
+- The names are changed from `SHOUT_CASE` to `PascalCase` to match C# style conventions.
+- There are a few special enum values, like `All`, that have been added to support cases in C where you’d have written a fixed constant like `0`.
```

Comment:
I think we want to mention that `All` contains all of the enums, so if you don't find the enum you are looking for in a group you should open an issue and use the `All` enum as a workaround.

Reply:
This took a bit of a rewrite; I hadn't looked closely at what `All` was originally, and erroneously assumed it was something it wasn't.  Here's the revised text:

> - The constants are grouped into strong `enum` types like `ErrorCode` and `ShaderType` — they’re not just a big pile of unrelated values like they are in C.
> - The names are changed from `SHOUT_CASE` to `PascalCase` to match C# style conventions.
> - The constants' `GL_` prefixes have been removed for readability.
>
> There is also a special enum type, `All`, that contains *all* of the GL constants; this can simplify porting code from other languages, and it ensures that all of the OpenGL constants are available.  `All` should be avoided in favor of more specialized enum types where possible.

Reply:
I think we should have something that mentions how you can cast an `All` enum to the type of enum you need. As that is apparently something that a lot of people don't really know that you can do, and it would be especially nice if we mentioned that.
For example see the workaround in https://github.com/opentk/opentk/issues/688
`GL.GetFloat((GetPName)All.MaxTextureMaxAnisotropy);`

These are actually relevant to a few well used enums, specifically anisotropic filtering texture parameter and some `SizedInternalFormats` that are missing (see https://github.com/opentk/opentk/issues/752)

Reply:
I added a short parenthetical about casting `All`:

> There is also a special enum type, `All`, that contains *all* of the GL constants; this can simplify porting code from other languages, and it ensures that all of the OpenGL constants are available.  `All` should be avoided in favor of more specialized enum types where possible (but it isn't *always* possible, so don't forget you can cast from `All` to `int` or to the enum type you need!).